### ISSUE TITLE
Redesing of BiDirectional

### ISSUE THREAD

**Issue by torressa** [torressa/cspy#43]
Given there's a fair few extensions to make, I think it's worth changing the structure a bit so that working on them is a bit easier.

Improvements/Features/Bugs include:

- Dominance and label elimination (discussed in #37)
- Parallelisation (also discussed in #37)
- `min_res` bug (raised in #41)
- Time limit parameter (discussed in #42)

The current structure doesn't lend itself that well to all of the above, therefore, I propose to split the current `BiDirectional` object into two. 
A `Search` helper object (for both backward and forward searches) and a modified `BiDirectional` object which will just orchestrate the searches and output the results.

## Helper `Search` object

Purposes:

- Advancing the search
- Dominance checking 
- Label elimination 
  - [x] Elementary and non-elementary cases #37 #52 
  - [x] Label feasibility #41

Individual Attributes (not shared)
- `direction` - direction of search
- `unprocessed_labels` nested dict with labels to help with label elimination.

## Modified `BiDirectional` object

Purposes:
- Orchestrating the search
  - [x] In the original way (`.run` method)
  - [ ] In parallel (`.run_parallel` method) #37 
  - [x] Stopping the search (time limit) #42 
- Joining final labels
   - [x] Label feasibility #41
   - [x] Stopping early (path cost under an input threshold) #37

Individual Attributes:
- Resource bounds (shared for parallel search)
- best labels
- final label (to output the results)

**Comment by torressa** [comment]
Initial commit in branch `patch43`.
Structure mostly set, just need to fix the interactions with the helper object.

**Comment by torressa** [comment]
Things ticked above are nearly done but need further testing:
- time limit
- threshold
- label elimination

**Comment by torressa** [comment]
Also, there should probably be a base object since there are a number of attributes that are shared between the the helper object and `BiDirectional`

**Comment by torressa** [comment]
Latest commit fixes unit tests! Parallel needs more testing

**Comment by torressa** [comment]
Elementary version doesn't work for the vrpy solomon instances.

**Comment by torressa** [comment]
Merge into dev as unittests passing, still needs work.

**Comment by torressa** [comment]
Works now on the solomon instances, like 10x slower than the old release...

**Comment by torressa** [comment]
https://github.com/Kuifje02/vrpy/issues/9#issuecomment-654394619

**Comment by torressa** [comment]
New work in branch `patch43-v0.1.1`. Reverted back to the latest release due to the slowness of the structure outline in the 
description of this issue. Will carry on with further speed up techniques to the future C++ version.
There are some label pruning techniques that are really easy to implement and will provide an improvement (happens in python too if the problem is big enough); for instance see [Lozano et al (2015)](https://pubsonline.informs.org/doi/abs/10.1287/trsc.2014.0582) and [Feillet et al (2004)](https://onlinelibrary.wiley.com/doi/abs/10.1002/net.20033). 

Will have to think about applying pruning techniques outside the algorithm, therefore allowing for other labelling algorithms to be easily implemented.


Time limit :heavy_check_mark: 
Parallel :x: (deferred) 
Label elimination :x: (deferred)

**Comment by torressa** [comment]
Time limit going over quite significantly for large problems.
Will release after fixing

**Comment by torressa** [comment]
Just committed work on the [cpp branch](https://github.com/torressa/cspy/tree/cpp)

Remains to be done:
- [ ] python packaging.

in `tools/dev/` running `./build -p -r` should create the python package and run the unit tests, however the dlls in `.libs` are not found by the python package.
- [ ] ci
- [ ] benchmarks with `direction = both`
- [ ] add more [benchmarks](https://github.com/lyandut/MyPulseForCSP/tree/master/Instance)
- [ ] use efficient labels (`best_labels`) as well as unprocessed to the dominance checks

**Comment by torressa** [comment]
Callbacks broken (issue32) as well as some of the python unittests (issue20, issue22, issue38 and issue41)

**Comment by torressa** [comment]
Python unittests working on [ubuntu](https://github.com/torressa/cspy/actions?query=workflow%3A%22Ubuntu+Python+tests%22)!
Once benchmarks are working we'll be good to go

**Comment by torressa** [comment]
Finally got it working on the solomon instances! @Kuifje02
![old_vs_new](https://user-images.githubusercontent.com/23246013/104640068-70174600-56a8-11eb-93e0-fd5f516a382f.png)
